engine/src/state_chain_observer/client.rs

```diff
@@ -265,66 +277,80 @@ impl<BaseRpcClient: base_rpc_api::BaseRpcApi + Send + Sync + 'static, SignedExtr
 				}
 			});
 
-		let incompatible_block_message = |block_version: SemVer, block: BlockInfo| {
-			lazy_format::lazy_format!(
-				"This version '{}' is incompatible with the current release '{block_version}' at block {}: {:?}.",
-				*CFE_VERSION,
-				block.number,
-				block.hash,
-			)
-		};
-
 		let mut processed_stream = {
 			let latest_block: BlockInfo = block_stream.next().await.unwrap()?;
+
 			let mut block_stream =
 				process_stream_fn(Box::pin(block_stream.make_try_cached(latest_block))).await?;
 
-			let latest_block = *block_stream.cache();
-			if wait_for_required_version {
-				let incompatible_blocks =
-					futures::stream::once(futures::future::ready(Ok::<_, anyhow::Error>(
-						latest_block,
-					)))
-					.chain(block_stream.by_ref())
-					.try_take_while(|block_info| {
-						let base_rpc_client = base_rpc_client.clone();
-						let block_info = *block_info;
-						async move {
-							Ok(
-								if let Err(block_version) = base_rpc_client
-									.check_block_compatibility(block_info.hash)
-									.await?
-								{
-									info!(
-										"{} WAITING for a compatible release version.",
-										incompatible_block_message(block_version, block_info)
-									);
-									true
-								} else {
-									false
-								},
-							)
-						}
-					})
-					.boxed();
-
-				// Note underlying stream ends in Error, therefore it is guaranteed try_for_each
-				// will not exit successfully until a compatible block is found
-				incompatible_blocks
-					.try_for_each(move |_| futures::future::ready(Ok::<_, anyhow::Error>(())))
-					.await?;
-			} else {
-				base_rpc_client.check_block_compatibility(latest_block.hash).await?.map_err(
-					|latest_block_version| {
-						anyhow::anyhow!(
-							"{}",
-							incompatible_block_message(latest_block_version, latest_block)
+			let first_block = *block_stream.cache();
+
+			let block_compatibility =
+				base_rpc_client.check_block_compatibility(first_block).await?;
+
+			match block_compatibility.compatibility {
```

Comment:
Sorry I believe there is another problem, if the engine starts between the upgrade, i.e. the finalized block is the old version, and the unfinalized is the new version. The old engine will creation of the unfinalized stream will error here. And incorrectly state the start_from block.

Reply:
I think the unfinalized case needs to wiat like to didn't before, instead of erroring.

Reply:
You could use the exist bool `error_on_incompatible_block`

Reply:
Seems like a simple fix

Reply:
Probably makes sense for both the NotYet, and NoLonger cases?

Reply:
What do you mean? We don't have a start_from for the unfinalised?